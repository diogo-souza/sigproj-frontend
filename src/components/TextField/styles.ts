import styled, { css, DefaultTheme } from 'styled-components';
import { TextFieldProps } from '.';

type IconPositionProps = Pick<TextFieldProps, 'iconPosition' | 'icon'>;

type ContainerProps = Pick<TextFieldProps, 'disabled' | 'readOnly'> & {
  error?: boolean;
};

export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition, icon }) => css`
    color: ${theme.colors.titleBlack};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.semiBold};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small} ${theme.colors.white}
        inset;
      filter: none;
    }

    ${!!icon &&
    css`
        padding-${iconPosition}: ${theme.spacings.xsmall};
    `}
  `}
`;

export const InputContainer = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    background: ${theme.colors.white};
    border-radius: ${theme.border.radius};
    padding: 0 12px;
    border: 0.1rem solid;
    border-color: ${theme.colors.gray};
    height: ${theme.spacings.large};
    &:hover {
      border-color: ${theme.colors.darkGray};
    }
    transition: border-color ${theme.transition.fast},
      box-shadow ${theme.transition.fast};
    &:focus-within {
      border-color: ${theme.colors.focus};
      box-shadow: 0 0 0 0.3rem ${theme.colors.focusShadow};
    }
  `}
`;

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    width: 2.2rem;
    color: ${theme.colors.gray};
    order: ${iconPosition === 'right' ? 1 : 0};
    & > svg {
      width: 100%;
      color: ${theme.colors.titleBlack};
    }
  `}
`;

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.danger};
    font-size: ${theme.font.sizes.small};
  `}
`;

const containerModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.darkGray};
      &::placeholder {
        color: currentColor;
      }
    }
    ${InputContainer} {
      background: ${theme.colors.lightGray};
    }
  `,
  error: (theme: DefaultTheme) => css`
    ${InputContainer} {
      border-color: ${theme.colors.danger};
      &:focus-within {
        box-shadow: 0 0 0 0.3rem ${theme.colors.dangerShadow};
      }
    }
    ${Icon} {
      color: ${theme.colors.danger};
    }
  `,
  readOnly: (theme: DefaultTheme) => css`
    ${InputContainer} {
      border-color: ${theme.colors.darkGray};
      background-color: ${theme.colors.lightGray};
      &:focus-within {
        border-color: ${theme.colors.focus};
        box-shadow: 0 0 0 0.3rem ${theme.colors.focusShadow};
      }
    }
  `,
};

export const Container = styled.div<ContainerProps>`
  ${({ theme, disabled, error, readOnly }) => css`
    ${disabled && containerModifiers.disabled(theme)}
    ${error && containerModifiers.error(theme)}
    ${!!readOnly && containerModifiers.readOnly(theme)}
  `}
`;
