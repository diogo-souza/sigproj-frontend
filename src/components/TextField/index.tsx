import React, { InputHTMLAttributes, useState } from 'react';

import { cep, cpf, phoneNumber, telephoneNumber } from 'utils/masks';

import Label from 'components/Label';
import * as S from './styles';

export type TextFieldProps = {
  onInputChange?: (value: string) => void;
  label?: string;
  initialValue?: string;
  icon?: React.ReactNode;
  iconPosition?: 'left' | 'right';
  disabled?: boolean;
  error?: string;
  mask?: 'cep' | 'cpf' | 'phoneNumber' | 'telephoneNumber';
} & InputHTMLAttributes<HTMLInputElement>;

const handleMask = {
  cep: (e: React.ChangeEvent<HTMLInputElement>) => cep(e),
  cpf: (e: React.ChangeEvent<HTMLInputElement>) => cpf(e),
  phoneNumber: (e: React.ChangeEvent<HTMLInputElement>) => phoneNumber(e),
  telephoneNumber: (e: React.ChangeEvent<HTMLInputElement>) =>
    telephoneNumber(e),
};

const TextField: React.FC<TextFieldProps> = ({
  name,
  label,
  icon,
  iconPosition = 'left',
  initialValue = '',
  disabled = false,
  error,
  onInputChange,
  style,
  className,
  readOnly,
  mask,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue || '');

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (mask) {
      const currentMask = handleMask[mask];
      currentMask(e);
    }
    const newValue = e.currentTarget.value;
    setValue(newValue);

    // eslint-disable-next-line no-unused-expressions
    !!onInputChange && onInputChange(newValue);
  };

  return (
    <S.Container
      disabled={disabled}
      error={!!error}
      style={style}
      className={className}
      readOnly={readOnly}
    >
      {!!label && (
        <Label htmlFor={name} disabled={disabled} hasError={!!error}>
          {label}
        </Label>
      )}
      <S.InputContainer>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          name={name}
          value={value}
          onChange={onChange}
          iconPosition={iconPosition}
          disabled={disabled}
          icon={icon}
          readOnly={readOnly}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputContainer>
      {!!error && <S.Error role="alert">{error}</S.Error>}
    </S.Container>
  );
};
export default TextField;
