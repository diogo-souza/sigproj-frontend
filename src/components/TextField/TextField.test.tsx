import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FaEnvelope as MailIcon } from 'react-icons/fa';

import { renderWithTheme } from 'utils/tests/helpers';

import TextField from '.';

describe('<TextField />', () => {
  it('should render with Label', () => {
    renderWithTheme(<TextField label="Label" name="Field" />);

    expect(screen.getByLabelText('Label')).toBeInTheDocument();
  });

  it('should render without Label', () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument();
  });

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder="hey you" />);

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument();
  });

  it('should renders with Icon on the right side', () => {
    renderWithTheme(
      <TextField icon={<MailIcon data-testid="icon" />} iconPosition="right" />,
    );

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({ order: 1 });
  });

  it('should chenges its value when typing', async () => {
    renderWithTheme(<TextField label="TextField" name="textfield" />);

    const input = screen.getByRole('textbox');
    const text = 'This is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
    });
  });

  it('should not changes its value when disabled', async () => {
    renderWithTheme(<TextField label="TextField" name="TextField" disabled />);

    const input = screen.getByRole('textbox');
    expect(input).toBeDisabled();

    const text = 'This is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
  });

  it('should render with Icon', () => {
    renderWithTheme(<TextField icon={<MailIcon data-testid="icon" />} />);

    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('should render with error', () => {
    const { container } = renderWithTheme(
      <TextField
        icon={<MailIcon data-testid="icon" />}
        label="TextField"
        name="TextField"
        error="Error message"
      />,
    );

    expect(screen.getByText('Error message')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be accessible by tab', () => {
    renderWithTheme(<TextField label="TextField" name="TextField" />);

    const input = screen.getByLabelText('TextField');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });

  it('should not accessible by tab when disabled', () => {
    renderWithTheme(<TextField label="TextField" name="TextField" disabled />);

    const input = screen.getByLabelText('TextField');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).not.toHaveFocus();
  });
});
