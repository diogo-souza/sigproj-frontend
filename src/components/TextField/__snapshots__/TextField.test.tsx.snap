// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render with error 1`] = `
.c2 {
  display: inline-block;
  font-size: 1.6rem;
  color: #171923;
  font-weight: 600;
  cursor: pointer;
  margin-bottom: 0.8rem;
}

.c7 {
  color: #171923;
  font-family: Open-Sans,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  font-weight: 500;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #E1E1E1;
  border-radius: 0.4rem;
  padding: 0 1.6rem;
  border: 0.1rem solid;
  border-color: #E1E1E1;
}

.c4:hover {
  border-color: #939393;
}

.c4:focus-within {
  border-color: #0d6efd;
  box-shadow: 0 0 0 0.2rem rgba(75,174,227,0.4);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2.2rem;
  color: #939393;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: 100%;
}

.c8 {
  color: #dc3545;
  font-size: 1.4rem;
}

.c0 .c3 {
  border-color: #dc3545;
}

.c0 .c3:focus-within {
  box-shadow: 0 0 0 0.2rem rgba(220,53,69,0.4);
}

.c0 .c5,
.c0 .c1 {
  color: #dc3545;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        data-testid="icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M502.3 190.8c3.9-3.1 9.7-.2 9.7 4.7V400c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V195.6c0-5 5.7-7.8 9.7-4.7 22.4 17.4 52.1 39.5 154.1 113.6 21.1 15.4 56.7 47.8 92.2 47.6 35.7.3 72-32.8 92.3-47.6 102-74.1 131.6-96.3 154-113.7zM256 320c23.2.4 56.6-29.2 73.4-41.4 132.7-96.3 142.8-104.7 173.4-128.7 5.8-4.5 9.2-11.5 9.2-18.9v-19c0-26.5-21.5-48-48-48H48C21.5 64 0 85.5 0 112v19c0 7.4 3.4 14.3 9.2 18.9 30.6 23.9 40.7 32.4 173.4 128.7 16.8 12.2 50.2 41.8 73.4 41.4z"
        />
      </svg>
    </div>
    <input
      class="c7"
      id="TextField"
      value=""
    />
  </div>
  <p
    class="c8"
  >
    Error message
  </p>
</div>
`;
